### ISSUE TITLE
Add vectorization to the par_vec (aka par_unseq) implementations of the parallel algorithms

### ISSUE THREAD

**Issue by brycelelbach** [STEllAR-GROUP/hpx#2271]
The par_vec (aka par_unseq) policy allows interleaving of element access functions, e.g. it is safe to the iterations of the algorithm.

Explicit engagement of compiler vectorizers through pragmas is probably the best way to ensure this occurs (e.g. `#pragma simd`, `#pragma omp simd`).

I will probably take a look into doing this myself while preparing my CppCon talk on parallel algorithms.

**Comment by diehlpk** [comment]
@brycelelbach @hkaiser Could you please add a project description here https://github.com/STEllAR-GROUP/hpx/wiki/GSoC-2017-Project-Ideas

**Comment by Johan511** [comment]
I am interested in working on this project. I have seen that in the previous PRs we have added openMP pragmas for vectorization and parallelisation of a loop. Can someone guide me on how I can start out with working on this issue?

**Comment by hkaiser** [comment]
> I am interested in working on this project. I have seen that in the previous PRs we have added openMP pragmas for vectorization and parallelisation of a loop. Can someone guide me on how I can start out with working on this issue?

Yes, we have implemented this for the first batch of algorithms. There are still algorithms left that have not been touched, though. Also, we would need a thorough performance analysis of the existing implementation, combined with improvements, if needed.

**Comment by Johan511** [comment]
par_unseq implementation for algorithms

* [x]  `adjacent_difference`
* [ ]  `inner_product`  does it support any execution policy, could not find doc. Do we Implement using transform reduce?
* [x]  `adjacent_find` 
* [x]  `all_of` `any_of` `none_of` 
------------------------------------
* [ ]  `copy` `copy_if` `copy_n`
* [ ]  `move`
* [ ]  `count` `count_if`
* [ ]  `equal` `mismatch` 
* [ ]  `exclusive_scan` `inclusive_scan` 
* [ ]  `reduce` `transform`
* [ ]  `fill` `fill_n`
* [ ]  `find` `find_end` `find_first_of` `find_if` `find_if_not`
* [x]  `for_each` `for_each_n`
* [ ]  `generate` `generate_n`
* [ ]  `is_heap` `is_heap_until` 
* [ ]  `is_partitioned` `is_sorted` `is_sorted_until`
* [ ]  `lexicographical_compare`
* [ ]  `max_element` `min_element` `minmax_element`
* [ ]  `make_heap`
* [ ]  `partial_sort`
* [ ]  `partial_sort_copy` `nth_element`
* [ ]  `sort` 
* [ ]  `stable_sort` 
* [ ]  `partition`
* [ ]  `partition_copy` 
* [ ]  `stable_partition`  
* [ ]  `remove` `remove_if` 
* [ ]  `remove_copy` `remove_copy_if` 
* [ ]  `replace` `replace_copy` `replace_copy_if` `replace_if`
* [ ]  `reverse` `reverse_copy` 
* [ ]  `rotate` `rotate_copy` 
* [ ]  `search` `search_n`
* [ ]  `set_difference` `set_intersection` `set_symmetric_difference` `set_union` `includes` 
* [ ]  `inplace_merge` 
* [ ]  `merge` 
* [ ]  `swap_ranges`
* [ ]  `uninitialized_copy` `uninitialized_copy_n`
* [ ]  `uninitialized_fill` `uninitialized_fill_n`
* [ ]  `uninitialized_default_construct` `uninitialized_default_construct_n`
* [ ]  `uninitialized_value_construct` `uninitialized_value_construct_n`
* [ ]  `uninitialized_move` `uninitialized_move_n`
* [ ]  `destroy` `destroy_n`
* [ ]  `unique`
* [ ]  `unique_copy`
* [ ]  `transform_reduce`
* [ ]  `transform_exclusive_scan` `transform_inclusive_scan` 
* [ ]  `shift_left` `shift_right` 
* [ ]  `starts_with` `ends_with`

**Comment by trkk28097402** [comment]
Hello @hkaiser , I am interest in this topic on gsoc24 ,I have a qeustion.
Is this restricted to only use the #pragma omp simd to vectorize or using something like __m128d, __m256d, some SIMD instructions are unreadable.

**Comment by hkaiser** [comment]
> Hello @hkaiser , I am interest in this topic on gsoc24 ,I have a qeustion. Is this restricted to only use the #pragma omp simd to vectorize or using something like __m128d, __m256d, some SIMD instructions are unreadable.

Everything is possible, I guess - as long as it is portable across architectures (beyond x86), at least in the long run.

**Comment by adhithyaragavan** [comment]
hi @hkaiser , i'm interested in picking this up. based on the previous threads it looks like there are still algorithms without par_unseq support would it make sense to start there, or is the priority more on the performance analysis + improvements to the existing implementations?
also happy to do both if there's a preferred order to tackle them in, would love to be guided on this

**Comment by adhithyaragavan** [comment]
Hi @hkaiser, 

I've been looking into picking up this issue and wanted to provide a structured status update based on a static routing audit of the current algorithms subsystem. 

Since the tracking checklist from earlier threads is somewhat stale, I traced the entry points, template overrides, and type dispatch mechanisms under `libs/core/algorithms` to see how execution policies route down to the `datapar` backend primitives (like those in `datapar/loop.hpp` or dedicated headers).

Because vectorization in HPX is explicit SIMD (`vector_pack_type_t<T>`) rather than relying purely on compiler auto-vectorization, I’ve categorized the current state of the ~70 algorithms into three distinct tiers:

### 1. Tier 1: Explicitly Vectorized / Capable
These algorithms successfully route to dedicated `datapar/*.hpp` drivers or delegate directly to other vectorized templates. 
* **Algorithms:** `count`, `count_if`, `find`, `find_if`, `find_if_not`, `all_of`, `any_of`, `none_of`, `equal`, `mismatch`, `fill`, `fill_n`, `replace`, `replace_if`, `search`, `transform`, `copy`, `copy_n`, `move` (via `transfer.hpp`).
* *Note:* For example, `count_if`'s internal functor layout accurately leverages horizontal mask reductions via `traits::count_bits(op_(proj_(*curr)))`.

### 2. Tier 2: Routes to Datapar Overrides (Empirical Verification Required)
These route through generic loop primitives (like `loop_ind` or `const_loop_n`) that support data-parallel execution policies. However, because type constraints can trigger silent scalar fallbacks without causing compilation errors, these need empirical isolation (IR/assembly probing) to guarantee they aren't executing scalar loops under the hood.
* **Algorithms:** `for_each`, `for_each_n`, `min_element`, `max_element`, `minmax_element`, `is_heap`, `is_heap_until`, `is_sorted`, `is_sorted_until`, `is_partitioned`, `lexicographical_compare`, `partition`, `remove`, `unique`, `destroy`, `reverse`.

### 3. Tier 3: Gaps (Genuine Scalar Fallbacks or Complex Structural Gaps)
These either drop completely straight down to a scalar engine or possess loop-carried dependencies (like carry propagation in scans) that limit SIMD efficiency.
* **Structural/Scan Gaps:** `inclusive_scan`, `exclusive_scan`, `transform_inclusive_scan`, `transform_exclusive_scan`, `transform_reduce` (binary), `reduce_by_key`.
* **Element-wise / Low-Hanging Fruit Gaps:** `iota`, `shift_left`, `shift_right`, `rotate`, `remove_copy`, and the `uninitialized_*` allocation family.

### Next Steps:
My local environment is an arm64 host, so I am setting up an empirical loop-vectorization check using standard compiler optimization diagnostics paired with direct NEON assembly inspection (`extern "C"` non-inlined probes) to clear the Tier 2 unknowns. 

Would it make sense to post those empirical findings here once done, or should I pivot immediately toward providing a fresh `datapar` implementation wrapper for one of the straightforward element-wise gaps in Tier 3 (like `iota` or `shift_left`)? Would love your guidance on the priority!

**Comment by hkaiser** [comment]
Please be our guest. We can use any help you may want to provide.

> Because vectorization in HPX is explicit SIMD (`vector_pack_type_t<T>`) rather than relying purely on compiler auto-vectorization, I’ve categorized the current state of the ~70 algorithms into three distinct tiers:

Just as a side note: the explicit SIMD algorithms you're referring to are used with the `hpx::execution::simd` and `hpx::execution::par_simd` execution policies. Compiler-based vectorization happens with the `hpx::execution::unseq` and `hpx::execution::par_unseq` policies. Those rely on different implementations.
